### ISSUE TITLE
[SNative] Don't return nil from SNative Dispatch

### ISSUE THREAD

**Issue by jasonklein** [hyperledger/burrow#418]
`permissionsContract` in https://github.com/eris-ltd/eris-db/blob/master/manager/eris-mint/evm/snative.go#L65 returns `nil` (among other return values) if the `caller` does not have the invoked permission.

The Solidity functions that are used to access the sNatives functions (see `Permissions`, https://github.com/eris-ltd/eris-pm/blob/master/tests/fixtures/app33-evm_connection_edb_permissions_and_roles_layer/permissions.sol) expect to receive `int`s (ideally changed to `uint64`s, see https://github.com/eris-ltd/eris-db/issues/417) or `bool`, depending on the function. Receiving `nil` gets `Invalid jump dest`.

Can `permissionsContract` be changed to return the default value of the permission intended to be invoked (so, either `0` or `false`, I think)?

**Comment by silasdavis** [comment]
I'll change the default return type to 0, which is both false and 0 in the EVM, and as you say the 0 flag grants no permissions, so this would seem to make sense, see: https://github.com/eris-ltd/eris-db/pull/500#issuecomment-280632471

**Comment by jasonklein** [comment]
@silasdavis, `nil` is also returned if the account the caller is referencing, e.g., if calling `set_base`, the account for which a permission is trying to be set, does not exist. Presumably this would also be resolved by the above change, but want to note it in case not.

**Comment by silasdavis** [comment]
@jasonklein (yay I managed to @ you) I've actually stepped back from this change. In all the cases you have cited nil (equivalent to empty byte slice) is returned when an error is also. This is handled by the vm here: https://github.com/eris-ltd/eris-db/blob/release-0.16/manager/eris-mint/evm/vm.go#L885

If an error is returned then the vm pushes `Zero256` on to the stack. So I don't think the nil returns can be the source of the invalid jump dest (I think @VoR0220 had another theory?). It is idiomatic to return nil from the go functions so I have reinstated these. Returning 0 at these functions may lead to a false reliance. When an error is returned it must be handled. In effect the behaviour that I thought I was adding is actually already in place via the EVM.

I think they invalid jump dest must have another cause. It might be worth re-checking this against current release-0.16 branch. We have made breaking changes (some along the lines of your suggestion) to the snatives.
